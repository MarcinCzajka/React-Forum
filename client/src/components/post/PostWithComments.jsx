import React from 'react';
import { Helmet } from "react-helmet";
import { Comment, Grid, Segment } from "semantic-ui-react";
import ForumPostContainer from './forumPost/ForumPostContainer';
import ForumPostPlaceholder from '../placeholders/ForumPostPlaceholder';
import CommentGroup from './comments/CommentGroup';
import getForumPost from './businessLogic/forumPostApi';
import './PostWithComments.css';

class PostWithComments extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            roomDetails: {},
            postDetailsReady: false
        }
    }

    componentDidMount() {
        //Pull id from URI and get ForumPost details
        getForumPost(this.props.match.params.id)
            .then(res => {
                this.setState({roomDetails: res, postDetailsReady: true})
            })
            .catch(err => {
                console.log(err);
            })
    }

    render() {
        const title = `${(this.state.roomDetails.title ? this.state.roomDetails.title + ' - ' : '')}React-forum`;
        
        return (
            <>
                <Helmet>
                    <title>{title}</title>
                </Helmet>

                <Grid centered className='roomGrid'>
                    <Comment.Group className='postGroupContainer'>

                        {this.state.postDetailsReady ? (
                            <ForumPostContainer
                                {...this.state.roomDetails}
                            />
                        ) : (
                            <div className='roomContainer'>
                                <ForumPostPlaceholder /> 
                            </div>
                        )}

                        <Segment className='initialPost'>
                            <CommentGroup
                                sorting={'{"date":"-1"}'}
                                initialPost={true}
                                roomId={this.props.match.params.id}
                                parentId={this.props.match.params.id}
                            />
                        </Segment>
                        
                    </Comment.Group>
                </Grid>
            </>
        )
    }

}

export default PostWithComments;