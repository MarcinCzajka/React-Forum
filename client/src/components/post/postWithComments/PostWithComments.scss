.roomGrid {
    margin-top: 2vh !important;
    margin: 0;
}

.postGroupContainer {
    max-width: 100% !important;
}

.initialPost > div {
    left: 0 !important;
}