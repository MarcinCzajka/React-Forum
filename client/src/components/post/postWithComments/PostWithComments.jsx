import React from 'react';
import { Comment, Grid, Segment } from "semantic-ui-react";
import ForumPostContainer from '../forumPost/ForumPostContainer';
import NewComment from './newComment/NewComment';
import CommentGroup from '../comments/CommentGroup';
import UserContext from '../../../contexts/UserContext';
import { replyToForumPost } from '../forumPost/forumPostLogic/forumPostApi';
import './PostWithComments.scss';

class PostWithComments extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            forumPostId: this.props.match.params.id,
            forumPostReady: false,
            replyContent: ''
        }

        this.forumPostRef = React.createRef();
    }

	static contextType = UserContext;

	handleReplyChange = (value) => {
		this.setState({replyContent: value});
	}
	
	handleReplyToPost = () => {
		replyToForumPost(this.context.userId, this.state.forumPostId, this.state.replyContent)
			.then(res => {
				if(res.status === 200) {
					this.refreshComments();
				}
			})
	}

    refreshComments = () => {
        window.location.reload();
    }

    removeForumPost = () => {
        //Function to be used in case Image is no longer available
        window.location = window.location.origin;
    }

    render() {
        const { forumPostId, replyContent } = this.state;

        return (
            <Grid centered className='roomGrid'>
                <Comment.Group className='postGroupContainer'>

                    <ForumPostContainer
                        _id={forumPostId}
                        refreshComments={this.refreshComments}
                        removeForumPost={this.removeForumPost}
                        ref={this.forumPostRef}
                    />

                    <Segment className='initialPost neuromorphSegment'>

                        <NewComment
                            replyContent={replyContent}
                            userAvatar={this.context.userAvatar}
                            handleReplyChange={this.handleReplyChange}
                            handleReplyToPost={this.handleReplyToPost}
                        />

                        <CommentGroup
                            sorting={'{"date":"-1"}'}
                            initialPost={true}
                            roomId={forumPostId}
                            parentId={forumPostId}
                        />
                    </Segment>
                    
                </Comment.Group>
            </Grid>
        )
    }

}

export default PostWithComments;