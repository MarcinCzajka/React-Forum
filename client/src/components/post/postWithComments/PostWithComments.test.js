import React from 'react';
import { create } from "react-test-renderer";
import { shallow } from 'enzyme';
import PostWithComments from './PostWithComments';
import { LocaleProvider } from '../../../contexts/LocaleContext';
import testLocale from '../../../locale/JestLocale/dictionary.json';

//I get my ID from props.match.params.id (query) and it is required to be there
const id = {
    params: {
        id: 'abcde12345'
    }
}

describe('NewComment component', () => {
    test('renders without crashing', () => {
        shallow(<PostWithComments match={id} />);
    });

    test('match snapshot', () => {
        const component = create(<LocaleProvider value={testLocale}><PostWithComments match={id} /></LocaleProvider>);
        expect(component).toMatchSnapshot();
    });

});