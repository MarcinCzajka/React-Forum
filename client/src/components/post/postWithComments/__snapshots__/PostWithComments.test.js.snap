// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewComment component match snapshot 1`] = `
<div
  className="ui centered grid roomGrid"
>
  <div
    className="ui comments postGroupContainer"
  >
    <article
      className="roomContainer postContainerOverlay"
    >
      <div
        className="postPlaceholder"
      >
        <div
          className="ui placeholder"
        >
          <div
            className="image"
          />
        </div>
        <div
          className="ui placeholder"
        >
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
          <div
            className="line"
          />
        </div>
      </div>
      
      
    </article>
    <div
      className="ui segment initialPost neuromorphSegment"
    >
      
      <div
        className="ui large comments newCommentContainer"
        style={
          Object {
            "display": "",
          }
        }
      >
        <div
          className="comment comment"
        >
          <div
            className="avatarPlaceholderContainer sizePostComment"
          >
            <i
              aria-hidden="true"
              className="user icon avatarPlaceholder"
              onClick={[Function]}
            />
          </div>
          <div
            className="avatar avatar"
            onLoad={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
          />
          <div
            className="pseudoLoggedIn"
          />
          <textarea
            className=""
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Post new comment..."
            value=""
          />
          <div
            className="buttonContainer hiddenButtonContainer"
          >
            <button
              className="ui teal compact disabled right floated button"
              disabled={true}
              onClick={[Function]}
              tabIndex={-1}
            >
              Comment
            </button>
            <button
              className="ui basic compact right floated button"
              onClick={[Function]}
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
      <div
        className="ui comments recursiveComment"
      />
    </div>
  </div>
</div>
`;
